### DISCUSSION TITLE
Simulations blowing up

### DISCUSSION THREAD

**Discussion by TheFlying12** [ProjectPhysX/FluidX3D#133] [Q&A] [answered]
I am trying to run the basic simulations, and all of them start out fine, then about 2-3 seconds in the image blows up and all I get is a red cube showing on my screen. Is there anything I need to do in defines or setup differently?
<img width="1268" alt="Screenshot 2023-11-29 at 1 21 01 PM" src="https://github.com/ProjectPhysX/FluidX3D/assets/78625538/8ddb8023-25e0-4790-b440-b678601f0b42">
<img width="723" alt="Screenshot 2023-11-29 at 1 20 43 PM" src="https://github.com/ProjectPhysX/FluidX3D/assets/78625538/a343fd9c-53d4-448c-91db-95347d8dd491">

**Comment by ProjectPhysX** [comment, marked as answer]
Try with enabling the [`SUBGRID`](https://github.com/ProjectPhysX/FluidX3D/blob/master/src/defines.hpp#L24) extension!

**Reply by TheFlying12** [reply]
Thank you! 
As a quick follow-up: Is there anything I can do about the resolution of the visualization?

**Reply by ProjectPhysX** [reply]
You can try the regular [`INTERACTIVE_GRAPHICS`](https://github.com/ProjectPhysX/FluidX3D/blob/master/src/defines.hpp#L27) mode, not [`INTERACTIVE_GRAPHICS_ASCII`](https://github.com/ProjectPhysX/FluidX3D/blob/master/src/defines.hpp#L28) (if you're on Linux, see the compile instructions for X11 graphics [here](https://github.com/ProjectPhysX/FluidX3D/blob/master/DOCUMENTATION.md#linux)). If your system can only do ASCII graphics, then you can reduce font size and make the window fullscreen to get more pixels. Or you can only enable the headless [`GRAPHICS`](https://github.com/ProjectPhysX/FluidX3D/blob/master/src/defines.hpp#L29) mode (comment out the other two), and [render images to the hard drive](https://github.com/ProjectPhysX/FluidX3D/blob/master/DOCUMENTATION.md#video-rendering).

**Reply by marcc1229** [reply]
Piggybacking on this is that's okay.  I'm having the same issue with lower resolution than I think I should be able to run.  Running it headless in graphics mode helps a little bit but in both cases after allocating over a certain point (7500u in this case) it will hang for a bit then just output a black screen.  HWmonitor claims I'm only using 45 percent of vram as well.

**Reply by ProjectPhysX** [reply]
@marcc1229 What GPU are you using? Sometimes approaching very close to the physical VRAM capacity leads to issues, because Windows also occupies a few hundred MB:
- it might spill VRAM into CPU RAM (runs super slow then, and you see VRAM occupation oscillating in HWmonitor)
- it might refuse to allocate without throwing an error (then HWmonitor shows lower than expected VRAM usage)
- when you actually exceed VRAM capacity in allocations you get a proper error message

Note that the graphics part also needs some VRAM for the frame buffer and z-buffer, which is not counted in the 7500u.

**Reply by marcc1229** [reply]
Using a 3090, the initial screen shows the appropriate 24000mb of memory.  I'm thinking it's the second bullet point.  Every now and then hwmonitor will stop at different points.  At first I thought it was resource limiting in the vs studio debugger but disabling that didn't help.  Running the application outside of the debugger doesn't seem to help either.   I did get the error when setting a higher number than the vram capacity.  The limit in HWmonitor seems to be 26ish percent.  If I set it to a higher number hwmonitor will show 30-60 percent usage but the run still fails.

**Reply by ProjectPhysX** [reply]
@marcc1229 can you try with lower resolution, compare what it shows in the console output under 
```| Memory Usage    |                                CPU ... MB, GPU 1x ... MB |```
with what you set in the `resolution(...)` function, and repeat with higher resolution up to the point of failure? If this persists, please open a new issue and post the entire `main_setup()` function there, I'll try to replicate then!

**Reply by marcc1229** [reply]
I did end up sort of getting it working over 1/4 of total vram (seems to be an opencl thing) but only at 1/2 and full vram usage.  Works both when using resolution directly and the max vram use function.  If you want I can start another thread with all that information but it's no longer an issue for me.

**Reply by ProjectPhysX** [reply]
@marcc1229 please open an issue on this. You should be able to use almost 100% of VRAM on you Nvidia GPU. This seems some kind of driver problem. Need more details!
